src/ImageSharp/Formats/Bmp/BmpDecoderCore.cs

```diff
@@ -130,8 +131,40 @@ public BmpDecoderCore(Configuration configuration, IBmpDecoderOptions options)
         ///    <para><paramref name="stream"/> is null.</para>
         /// </exception>
         /// <returns>The decoded image.</returns>
-        public Image<TPixel> Decode<TPixel>(Stream stream)
+        public async Task<Image<TPixel>> DecodeAsync<TPixel>(Stream stream)
             where TPixel : unmanaged, IPixel<TPixel>
+        {
+            // if we can seek then we arn't in a context that errors on async operations
+            if (stream.CanSeek)
+            {
+                return this.Decode<TPixel>(stream);
+            }
+            else
+            {
+                // cheat for now do async copy of the stream into memory stream and use the sync version
```

Comment:
I use a pool backed memory stream that allocates small chunks. Not sure how performant reading would be though.

https://github.com/SixLabors/ImageSharp.Web/blob/master/src/ImageSharp.Web/ChunkedMemoryStream.cs

Reply:
Is maybe [RecyclableMemoryStream](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream) an option? Its meant to be a drop in replacement for `MemoryStream`

Reply:
I want to have a look at a read-only version of [this](https://github.com/mgravell/Pipelines.Sockets.Unofficial/compare/mgravell:24482d4...mgravell:6740ea4) less overhead and should support all out target frameworks.

Reply:
Haven't noticed the ongoing discussion and made a comment about this as well.
`RecyclableMemoryStream` seems more sophisticated at first glance. 

Whichever we choose, I think it would make sense to integrate our adapted solution with `MemoryAllocator`.